contracts/ERC721TokenRegistry.sol

```diff
@@ -0,0 +1,193 @@
+
+        if (attributes.tokenAddress == address(0)) {
+            // The token has not yet been added to the registry.
+            attributes.tokenAddress = _tokenAddress;
+            attributes.name = _tokenName;
+            attributes.tokenIndex = tokenSymbolListLength;
+
+            tokenSymbolList[tokenSymbolListLength] = _symbol;
+            tokenSymbolListLength++;
+        } else {
+            // The token has already been added to the registry; update attributes.
+            attributes.tokenAddress = _tokenAddress;
+            attributes.name = _tokenName;
+        }
+
+        // Update this contract's storage.
+        symbolHashToTokenAttributes[symbolHash] = attributes;
+    }
+
+    /**
+     * Given a symbol, resolves the current address of the token the symbol is mapped to.
+     */
+    function getTokenAddressBySymbol(string _symbol) public view returns (address) {
+        bytes32 symbolHash = keccak256(_symbol);
+
+        TokenAttributes storage attributes = symbolHashToTokenAttributes[symbolHash];
+
+        return attributes.tokenAddress;
+    }
+
+    /**
+     * Given the known index of a token within the registry's symbol list,
+     * returns the address of the token mapped to the symbol at that index.
+     */
+    function getTokenAddressByIndex(uint _index) public view returns (address) {
+        string storage symbol = tokenSymbolList[_index];
+
+        return getTokenAddressBySymbol(symbol);
+    }
+
+    /**
+     * Given a symbol, resolves the index of the token the symbol is mapped to within the registry's
+     * symbol list.
+     */
+    function getTokenIndexBySymbol(string _symbol) public view returns (uint) {
+        bytes32 symbolHash = keccak256(_symbol);
+
+        TokenAttributes storage attributes = symbolHashToTokenAttributes[symbolHash];
+
+        return attributes.tokenIndex;
+    }
+
+    /**
+     * Given an index, resolves the symbol of the token at that index in the registry's
+     * token symbol list.
+     */
+    function getTokenSymbolByIndex(uint _index) public view returns (string) {
+        return tokenSymbolList[_index];
+    }
+
+    /**
+     * Given a symbol, returns the name of the token the symbol is mapped to within the registry's
+     * symbol list.
+     */
+    function getTokenNameBySymbol(string _symbol) public view returns (string) {
+        bytes32 symbolHash = keccak256(_symbol);
+
+        TokenAttributes storage attributes = symbolHashToTokenAttributes[symbolHash];
+
+        return attributes.name;
+    }
+
+    /**
+     * Given the index for a token in the registry, returns the name of the token as provided in
+     * the associated TokensAttribute struct.
+     *
+     * Example:
+     *   getTokenNameByIndex(1);
+     *   => "CryptoKitties"
+     */
+    function getTokenNameByIndex(uint _index) public view returns (string) {
+        string memory symbol = getTokenSymbolByIndex(_index);
```

Comment:
so i'm confused as to how we can use memory here, yet need to use storage elsewhere.

Reply:
Well, for one I get a compilation error if I try to change it:

```
TypeError: Type string memory is not implicitly convertible to expected type string storage pointer.
        string storage symbol = getTokenSymbolByIndex(_index);
        ^---------------------------------------------------^
```

I think this is explained by:

> As references are passed internally through function parameters, remember they default to memory and if the variable was on storage it would create a copy and any modification would not persist.
- https://medium.com/coinmonks/ethereum-solidity-memory-vs-storage-which-to-use-in-local-functions-72b593c3703a